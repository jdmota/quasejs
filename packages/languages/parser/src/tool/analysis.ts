import { DState } from "./state";
import {
  Transition, RuleTransition, PredicateTransition,
  RangeTransition, EOFTransition, NamedTransition, TokenFinalTransition
} from "./transitions";
import { MapKeyToSet } from "./utils/map-key-to-set";
import { MapKeyToValue } from "./utils/map-key-to-value";
import { MapRangeToValue } from "./utils/map-range-to-value";
import { MapRangeToSet } from "./utils/map-range-to-set";
import { Rule, ParserRule, LexerRule } from "./parser/grammar-parser";

export class Context {

  parent: Context | null;
  size: number;

  constructor( parent: Context | null ) {
    this.parent = parent;
    this.size = parent ? parent.size + 1 : 0;
  }

}

export type Look = RangeTransition | PredicateTransition | EOFTransition;
export type GoTo = [ Transition, DState ] | null;

class LookSet {

  transitions: MapKeyToValue<PredicateTransition | EOFTransition, null>;
  ranges: MapRangeToValue<null>;

  constructor() {
    this.transitions = new MapKeyToValue();
    this.ranges = new MapRangeToValue();
  }

  add( transition: Look ) {
    if ( transition instanceof RangeTransition ) {
      this.ranges.addRange( transition.from, transition.to, null );
    } else {
      this.transitions.add( transition, null );
    }
  }

  importFrom( other: LookSet ) {
    for ( const [ transition, value ] of other.transitions ) {
      this.transitions.add( transition, value );
    }
    for ( const [ range, value ] of other.ranges ) {
      this.ranges.addRange( range.from, range.to, value );
    }
  }

  * [Symbol.iterator](): IterableIterator<Look> {
    for ( const [ transition ] of this.transitions ) {
      yield transition;
    }
    for ( const [ range ] of this.ranges ) {
      yield new RangeTransition( range.from, range.to );
    }
  }

}

class LookData {

  transitions: MapKeyToSet<PredicateTransition | EOFTransition, GoTo>;
  ranges: MapRangeToSet<GoTo>;

  constructor() {
    this.transitions = new MapKeyToSet();
    this.ranges = new MapRangeToSet();
  }

  add( transition: Look, value: GoTo ) {
    if ( transition instanceof RangeTransition ) {
      this.ranges.addRange( transition.from, transition.to, new Set( [ value ] ) );
    } else {
      this.transitions.addOne( transition, value );
    }
  }

  * [Symbol.iterator](): IterableIterator<[Look, Set<GoTo>]> {
    for ( const entry of this.transitions ) {
      yield entry;
    }
    for ( const [ range, value ] of this.ranges ) {
      yield [ new RangeTransition( range.from, range.to ), value ];
    }
  }

}

function sortConflicts( a: GoTo, b: GoTo ) {
  if ( a != null && a[ 0 ] instanceof TokenFinalTransition ) {
    if ( b != null && b[ 0 ] instanceof TokenFinalTransition ) {
      return a[ 0 ].id - b[ 0 ].id;
    }
    return 1;
  }
  return -1;
}

export class Analyser {

  initialStates: Map<Rule, DState>;
  finalStates: Map<DState, Rule | null>;
  follows: Map<Rule, Set<DState>>;
  lookahead: Map<DState, LookSet>;
  conflicts: string[];

  constructor( { initialStates, finalStates, follows }: {
    initialStates: Map<Rule, DState>;
    finalStates: Map<DState, Rule | null>;
    follows: Map<Rule, Set<DState>>;
  } ) {
    this.initialStates = initialStates;
    this.finalStates = finalStates;
    this.follows = follows;
    this.lookahead = new Map();
    this.conflicts = [];
  }

  lookaheadForState( state: DState ) {
    let set = this.lookahead.get( state );
    let existed = true;
    if ( !set ) {
      set = new LookSet();
      existed = false;
      this.lookahead.set( state, set );
    }
    return {
      set,
      existed
    };
  }

  testConflict( rule: ParserRule | LexerRule | null, state: DState, look: Look, set: Set<GoTo> ) {
    const arr = Array.from( set ).sort( sortConflicts );
    if ( arr.length > 1 ) {
      this.conflicts.push(
        `In ${rule ? `rule ${rule.name}` : "lexer"}, in state ${state.id}, when seeing ${look}, multiple choices: ` +
        `${arr.map( goto => ( goto ? `${goto[ 0 ]} to ${goto[ 1 ].id}` : "leave" ) ).join( "; " )}`
      );
    } else if ( arr.length === 0 ) {
      throw new Error( "Assertion error" );
    }
    return arr[ 0 ];
  }

  _analyseFinalState( state: DState, ctx: Context | null, set: LookSet ) {
    let ret = false;
    if ( ctx ) {
      if ( ctx.size === 0 ) {
        // EOF
        set.add( new EOFTransition() );
      }
      ret = true;
      return ret;
    }
    // Follow
    const rule = this.finalStates.get( state );
    const f = rule && this.follows.get( rule );
    if ( f && f.size > 0 ) {
      for ( const dest of f ) {
        const result = this._analyse( dest, ctx );
        set.importFrom( result.set );
        ret = result.ret || ret;
      }
      if ( rule && rule.modifiers.start ) {
        // EOF
        set.add( new EOFTransition() );
      }
    } else {
      // EOF
      set.add( new EOFTransition() );
    }
    return ret;
  }

  _analyse( state: DState, ctx: Context | null ): { set: LookSet; ret: boolean } {
    const { set, existed } = this.lookaheadForState( state );
    if ( existed ) {
      return {
        set,
        ret: false
      };
    }

    let ret = false;

    if ( this.finalStates.has( state ) ) {
      ret = this._analyseFinalState( state, ctx, set );
    }

    for ( const [ transition, dest ] of state ) {
      ret = this._analyseTransition( transition, dest, ctx, set ) || ret;
    }

    return {
      set,
      ret
    };
  }

  _analyseTransition(
    transition: Transition, dest: DState, ctx: Context | null, set: LookSet
  ): boolean {
    if ( transition instanceof RuleTransition ) {
      const newCtx = ctx ? new Context( ctx ) : null;

      const result = this._analyse( this.initialStates.get( transition.rule ) as DState, newCtx );
      set.importFrom( result.set );

      if ( result.ret ) {
        const result2 = this._analyse( dest, ctx );
        set.importFrom( result2.set );
        return result2.ret;
      }
      return false;
    }
    if ( transition instanceof PredicateTransition ) {
      set.add( transition );
      return false;
    }
    if ( transition instanceof RangeTransition ) {
      set.add( transition );
      return false;
    }
    if ( transition instanceof NamedTransition ) {
      return this._analyseTransition( transition.subTransition, dest, ctx, set );
    }
    if ( transition.isEpsilon ) {
      const result = this._analyse( dest, ctx );
      set.importFrom( result.set );
      return result.ret;
    }
    throw new Error( "Assertion error" );
  }

  analyse( state: DState, ctx: Context | null ) {

    const lookData: LookData = new LookData(); // look<A, B> -> if we see A, go to B

    if ( this.finalStates.has( state ) ) {
      const set: LookSet = new LookSet();
      this._analyseFinalState( state, ctx, set );
      for ( const look of set ) {
        lookData.add( look, null );
      }
    }

    for ( const [ transition, dest ] of state ) {
      const set: LookSet = new LookSet();
      this._analyseTransition( transition, dest, ctx, set );
      for ( const look of set ) {
        lookData.add( look, [ transition, dest ] );
      }
    }

    return lookData;
  }

}
