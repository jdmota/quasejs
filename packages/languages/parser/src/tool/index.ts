import Grammar, { Options } from "./grammar";
import { Factory } from "./factory";
import { NfaToDfa, DfaMinimizer } from "./optimizer";
import { Analyser } from "./analysis";
import { ParserRule, LexerRule } from "./parser/grammar-parser";
import { DFA } from "./abstract-optimizer";
import { DState } from "./state";
import { CodeGenerator } from "./code-gen";
import { Frag } from "./automaton";

export default function(grammarText: string, options?: Options) {
  // Init

  const grammar = new Grammar(grammarText, options);

  const parserRulesFactory = new Factory(grammar, false);
  const lexerRulesFactory = new Factory(grammar, true);

  const parserRuleToAutomaton: Map<ParserRule, DFA<DState>> = new Map();
  const lexerRuleToAutomaton: Map<LexerRule, DFA<DState>> = new Map();

  const nfaToDfa = new NfaToDfa();
  const dfaMinimzer = new DfaMinimizer();

  function minimize(frag: Frag) {
    return dfaMinimzer.minimize(
      nfaToDfa.do({
        start: frag.in,
        acceptingSet: new Set([frag.out]),
      })
    );
  }

  const analyzerOpts = {
    initialStates: new Map(),
    finalStates: new Map(),
    follows: dfaMinimzer.follows,
  };
  const analyser = new Analyser(analyzerOpts);

  // Create automaton for each rule and safe information about each

  function handleRule(frag: Frag, rule: ParserRule | LexerRule) {
    const minimized = minimize(frag);

    analyzerOpts.initialStates.set(rule, minimized.start);
    for (const state of minimized.acceptingSet) {
      analyzerOpts.finalStates.set(state, rule);
    }
    return minimized;
  }

  for (const rule of grammar.parserRules.values()) {
    const frag = parserRulesFactory.ParserRule(rule);
    parserRuleToAutomaton.set(rule, handleRule(frag, rule));
  }

  for (const rule of grammar.lexerRules.values()) {
    const frag = lexerRulesFactory.LexerRule(rule);
    lexerRuleToAutomaton.set(rule, handleRule(frag, rule));
  }

  // Create main lexer automaton

  const lexerAutomaton = minimize(lexerRulesFactory.genLexer(grammar.nodeToId));
  for (const state of lexerAutomaton.acceptingSet) {
    analyzerOpts.finalStates.set(state, null);
  }

  // Generate code

  const codeGen = new CodeGenerator(grammar, analyser);

  return {
    code: codeGen.gen(
      parserRuleToAutomaton,
      lexerRuleToAutomaton,
      lexerAutomaton
    ),
    conflicts: analyser.conflicts,
  };
}
