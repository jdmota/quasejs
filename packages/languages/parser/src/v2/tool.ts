import { Analyzer, AnalyzerFollow } from "./analysis/analysis";
import { Automaton, Frag } from "./automaton/automaton";
import { DState } from "./automaton/state";
import { FactoryRule } from "./factories/factory-rule";
import { FactoryToken } from "./factories/factory-token";
import { CfgToCode, CodeBlock } from "./generators/dfa-to-code/cfg-to-code";
import { ParserGenerator } from "./generators/generate-parser";
import { createGrammar } from "./grammar/grammar";
import {
  Declaration,
  RuleDeclaration,
  TokenDeclaration,
  TokenRules,
} from "./grammar/grammar-builder";
import { GrammarTypesInfer } from "./grammar/grammar-infer2";
import { DFA } from "./optimizer/abstract-optimizer";
import { DfaMinimizer, NfaToDfa } from "./optimizer/optimizer";

type ToolInput = {
  readonly name: string;
  readonly ruleDecls: readonly RuleDeclaration[];
  readonly tokenDecls: readonly TokenDeclaration[];
};

export function tool(opts: ToolInput) {
  const result = createGrammar(opts.name, opts.ruleDecls, opts.tokenDecls);

  if (result.errors) {
    for (const error of result.errors) {
      console.error(`Error in grammar ${opts.name}: ${error}`);
    }
    return null;
  }

  const grammar = result.grammar;
  const rulesAutomaton = new Automaton();
  const tokensAutomaton = new Automaton();

  // Init information the analyzer will need
  const initialStates = new Map<string, DState>();
  const follows = new Map<string, AnalyzerFollow[]>();

  // Init minimizers
  const nfaToDfa = new NfaToDfa();
  const dfaMinimizer = new DfaMinimizer(follows);

  function minimize(ruleName: string, frag: Frag) {
    dfaMinimizer.setCurrentRule(ruleName);
    return dfaMinimizer.minimize(
      nfaToDfa.do({
        start: frag.start,
        acceptingSet: new Set([frag.end]),
      })
    );
  }

  // Process rule declarations
  const ruleAutomatons = new Map<RuleDeclaration, DFA<DState>>();
  for (const rule of grammar.getRules()) {
    const frag = FactoryRule.process(grammar, rule, rulesAutomaton);
    const automaton = minimize(rule.name, frag);
    ruleAutomatons.set(rule, automaton);
    initialStates.set(rule.name, automaton.start);
  }

  // Process tokens
  const tokenAutomatons = new Map<TokenDeclaration, DFA<DState>>();
  for (const token of grammar.getTokens()) {
    const frag = FactoryToken.process(grammar, token, tokensAutomaton);
    const automaton = minimize(token.name, frag);
    tokenAutomatons.set(token, automaton);
    initialStates.set(token.name, automaton.start);
  }

  // Init analyzer
  const analyzer = new Analyzer({
    initialStates,
    follows,
  });

  // Create code blocks for tokens
  const tokenCodeBlocks = new Map<TokenDeclaration, CodeBlock>();
  for (const [token, automaton] of tokenAutomatons) {
    tokenCodeBlocks.set(token, new CfgToCode().process(automaton));
  }

  // Produce code for tokens
  const tokenCode = new Map<TokenDeclaration, string>();
  for (const [rule, block] of tokenCodeBlocks) {
    tokenCode.set(
      rule,
      new ParserGenerator(grammar, analyzer, rule).process(block)
    );
  }

  // Create code blocks for rules
  const ruleCodeBlocks = new Map<RuleDeclaration, CodeBlock>();
  for (const [rule, automaton] of ruleAutomatons) {
    ruleCodeBlocks.set(rule, new CfgToCode().process(automaton));
  }

  // Produce code for rules
  const ruleCode = new Map<RuleDeclaration, string>();
  for (const [rule, block] of ruleCodeBlocks) {
    ruleCode.set(
      rule,
      new ParserGenerator(grammar, analyzer, rule).process(block)
    );
  }

  const infer = new GrammarTypesInfer(grammar);
  for (const rule of grammar.getRules()) {
    infer.run(rule);
  }
  infer.debug();

  return { tokenCode, ruleCode };
}
