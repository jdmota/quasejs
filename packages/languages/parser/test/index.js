import tool from "../src/tool";

function runParser( generation, text ) {
  let result;
  /* eslint no-eval: 0 */
  eval( `
    ${generation.replace( "@quase/parser", "../src/runtime" )}\n
    result = new Parser(${JSON.stringify( text )}).parse();`
  );
  return result;
}

it( "basic", () => {

  const { code, conflicts } = tool(
    `
    @lexer
    FUN: 'fun';
    END: 'end';
    COMMA: ',';
    ARROW: '->';
    fragment ID_START: /[a-zA-Z]/;
    fragment ID_CHAR: /[a-zA-Z0-9]/;
    ID: ID_START ID_CHAR*;
    NUM: /[0-9]+/;

    @parser
    start PROGRAM: FUN ( params+=ID ( COMMA params+=ID )* )? ARROW body=EXP END;
    EXP: stuff=NUM | stuff=ID;
    `
  );

  expect( code ).toMatchSnapshot( "code" );
  expect( conflicts ).toMatchSnapshot( "conflicts" );

  expect( runParser( code, "fun id1, id2 -> 10000 end" ) ).toMatchSnapshot( "ast" );
  expect( runParser( code, "fun id1 -> id1 end" ) ).toMatchSnapshot( "ast" );
  expect( () => runParser( code, "fun id1 -> id1" ) ).toThrowErrorMatchingSnapshot( "error" );

} );

it( "supports empty", () => {

  const { code, conflicts } = tool(
    `
    start RULE1 : 'A' | ;
    `
  );

  expect( code ).toMatchSnapshot( "code" );
  expect( conflicts ).toMatchSnapshot( "conflicts" );

} );

it( "optimized repetitions and nested rule on right side", () => {

  const { code, conflicts } = tool(
    `
    start RULE1 : 'A'* 'A' RULE2;
    RULE2 : 'B'+ 'B';
    `
  );

  expect( code ).toMatchSnapshot( "code" );
  expect( conflicts ).toMatchSnapshot( "conflicts" );

} );

it( "negative character class in regexp", () => {

  const { code, conflicts } = tool(
    `
    @lexer
    STRING: /"([^\\\\"]|\\\\[^])*"/;

    @parser
    start RULE1 : STRING;
    `
  );

  expect( code ).toMatchSnapshot( "code" );
  expect( conflicts ).toMatchSnapshot( "conflicts" );

} );

it( "typescript", () => {

  const { code } = tool(
    `
    @lexer
    ID: /[a-zA-Z][a-zA-Z0-9]*/;
    NUM: /[0-9]+/;

    @parser
    start PROGRAM: 'fun' ( params+=ID ( ',' params+=ID )* )? '->' body=EXP 'end';
    EXP: stuff=NUM | stuff=ID;
    `,
    {
      typescript: true
    }
  );

  expect( code ).toMatchSnapshot( "code" );

} );

it( "actions", () => {

  const { code } = tool(
    `
    @lexer
    ID: /[a-zA-Z][a-zA-Z0-9]*/;
    NUM: /[0-9]+/;

    @parser
    start PROGRAM: ( id=ID { console.log("\\n",$id); if(true){} } )*;
    `
  );

  expect( code ).toMatchSnapshot( "code" );

} );
