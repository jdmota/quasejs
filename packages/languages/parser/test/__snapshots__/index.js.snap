// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic: ast 1`] = `
Object {
  "body": Object {
    "loc": Object {
      "end": Object {
        "column": 21,
        "line": 1,
        "pos": 21,
      },
      "start": Object {
        "column": 16,
        "line": 1,
        "pos": 16,
      },
    },
    "stuff": Object {
      "id": 6,
      "image": "10000",
      "label": "NUM",
    },
    "type": "EXP",
  },
  "loc": Object {
    "end": Object {
      "column": 25,
      "line": 1,
      "pos": 25,
    },
    "start": Object {
      "column": 0,
      "line": 1,
      "pos": 0,
    },
  },
  "params": Array [
    Object {
      "id": 5,
      "image": "id1",
      "label": "ID",
    },
    Object {
      "id": 5,
      "image": "id2",
      "label": "ID",
    },
  ],
  "type": "PROGRAM",
}
`;

exports[`basic: ast 2`] = `
Object {
  "body": Object {
    "loc": Object {
      "end": Object {
        "column": 14,
        "line": 1,
        "pos": 14,
      },
      "start": Object {
        "column": 11,
        "line": 1,
        "pos": 11,
      },
    },
    "stuff": Object {
      "id": 5,
      "image": "id1",
      "label": "ID",
    },
    "type": "EXP",
  },
  "loc": Object {
    "end": Object {
      "column": 18,
      "line": 1,
      "pos": 18,
    },
    "start": Object {
      "column": 0,
      "line": 1,
      "pos": 0,
    },
  },
  "params": Array [
    Object {
      "id": 5,
      "image": "id1",
      "label": "ID",
    },
  ],
  "type": "PROGRAM",
}
`;

exports[`basic: ast 3`] = `"Unexpected token eof, expected END (at 1:14)"`;

exports[`basic: conflicts 1`] = `Array []`;

exports[`basic: generation 1`] = `
"/* eslint-disable */
const Q=require(\\"@quase/parser\\");


  class Tokenizer extends Q.Tokenizer{
    
    constructor(input){
      super(input);
      this.table=[c=>0,c=>c==44?2:c==45?3:48<=c&&c<=57?5:65<=c&&c<=90?6:97<=c&&c<=100?6:c==101?7:c==102?10:103<=c&&c<=122?6:0,c=>0,c=>c==62?4:0,c=>0,c=>48<=c&&c<=57?5:0,c=>48<=c&&c<=57?6:65<=c&&c<=90?6:97<=c&&c<=122?6:0,c=>48<=c&&c<=57?6:65<=c&&c<=90?6:97<=c&&c<=109?6:c==110?8:111<=c&&c<=122?6:0,c=>48<=c&&c<=57?6:65<=c&&c<=90?6:97<=c&&c<=99?6:c==100?9:101<=c&&c<=122?6:0,c=>48<=c&&c<=57?6:65<=c&&c<=90?6:97<=c&&c<=122?6:0,c=>48<=c&&c<=57?6:65<=c&&c<=90?6:97<=c&&c<=116?6:c==117?11:118<=c&&c<=122?6:0,c=>48<=c&&c<=57?6:65<=c&&c<=90?6:97<=c&&c<=109?6:c==110?12:111<=c&&c<=122?6:0,c=>48<=c&&c<=57?6:65<=c&&c<=90?6:97<=c&&c<=122?6:0];
      this.finals={\\"2\\":3,\\"4\\":4,\\"5\\":6,\\"6\\":5,\\"7\\":5,\\"8\\":5,\\"9\\":2,\\"10\\":5,\\"11\\":5,\\"12\\":1};
      this.labels=[\\"\\",\\"FUN\\",\\"END\\",\\"COMMA\\",\\"ARROW\\",\\"ID\\",\\"NUM\\"];
    }
    initial(){return{id:\\"initial\\",label:\\"initial\\"};}
    eof(){return{id:\\"eof\\",label:\\"eof\\"};}
    
  codePointAt(index) {
    const first = this.input.charCodeAt(index);
    const size = this.input.length;
    if (
      first >= 0xD800 && first <= 0xDBFF &&
      size > index + 1
    ) {
      const second = this.input.charCodeAt(index + 1);
      if (second >= 0xDC00 && second <= 0xDFFF) {
        return (first - 0xD800) * 0x400 + second - 0xDC00 + 0x10000;
      }
    }
    return first;
  }
  
    
readToken() {
  const { input, table, finals, labels } = this;
  const length = input.length;
  const prevPos = this.pos;

  let state = 1;
  let i = this.pos;

  while ( i < length ) {
    const char = this.codePointAt( i );
    const nextState = table[ state ]( char );
    if ( nextState ) {
      state = nextState;
      if ( char > 0xffff ) {
        i += 2;
      } else {
        i++;
      }
    } else {
      break;
    }
  }

  const id = finals[state];
  if ( id == null ) {
    this.unexpectedChar();
  }
  const image=input.slice(prevPos,i);
  const splitted=image.split(/\\\\r\\\\n?|\\\\n/g);
  const newLines=splitted.length-1;
  this.pos=i;
  if (newLines>0) {
    this._lineStart=this.pos-splitted[newLines].length;
    this._curLine+=newLines;
  }
  return {
    id,
    label: labels[id],
    image
  };
}
  }
class Parser extends Q.Parser{
constructor(text){super(new Tokenizer(text));}

    expect(id) {
      const token = this.token;
      if ( token.id !== id ) {
        const labels = this.tokenizer.labels;
        throw this.error(
          \`Unexpected token \${labels[token.id]||token.id}, expected \${labels[id]||id}\`
        );
      }
      this.next();
      return token;
    }
    
/*EXP*/f2(){let $l=this.startNode();let $n_stuff;switch(this.token.id){case 6:$n_stuff=this.expect(6);break;case 5:$n_stuff=this.expect(5);break;default:this.unexpected();}return {type:\\"EXP\\",stuff:$n_stuff,loc:this.locNode($l)};}
/*PROGRAM*/f1(){let $l=this.startNode();let $n_params=[],$n_body;this.expect(1);switch(this.token.id){case 5:$n_params.push(this.expect(5));loop:while(true){switch(this.token.id){case 3:this.expect(3);$n_params.push(this.expect(5));break;default:break loop;}}break;}this.expect(4);$n_body=this.f2();this.expect(2);this.expect(\\"eof\\");return {type:\\"PROGRAM\\",params:$n_params,body:$n_body,loc:this.locNode($l)};}
parse(){return this.f1();}}
module.exports=Parser;
"
`;

exports[`check conflicts on repetitions: conflicts 1`] = `
Array [
  "In ZeroOrMore(2:18-2:22), when finding 'A', choice conflicts: 'A' (2:18-2:21) | Empty(2:18-2:22)",
  "In OneOrMore(3:12-3:16), when finding 'B', choice conflicts: 'B' (3:12-3:15) | Empty(3:12-3:16)",
]
`;

exports[`check conflicts on repetitions: generation 1`] = `
"/* eslint-disable */
const Q=require(\\"@quase/parser\\");


  class Tokenizer extends Q.Tokenizer{
    
    constructor(input){
      super(input);
      this.table=[c=>0,c=>c==65?2:c==66?3:0,c=>0,c=>0];
      this.finals={\\"2\\":1,\\"3\\":2};
      this.labels=[\\"\\",\\"'A'\\",\\"'B'\\"];
    }
    initial(){return{id:\\"initial\\",label:\\"initial\\"};}
    eof(){return{id:\\"eof\\",label:\\"eof\\"};}
    
  codePointAt(index) {
    const first = this.input.charCodeAt(index);
    const size = this.input.length;
    if (
      first >= 0xD800 && first <= 0xDBFF &&
      size > index + 1
    ) {
      const second = this.input.charCodeAt(index + 1);
      if (second >= 0xDC00 && second <= 0xDFFF) {
        return (first - 0xD800) * 0x400 + second - 0xDC00 + 0x10000;
      }
    }
    return first;
  }
  
    
readToken() {
  const { input, table, finals, labels } = this;
  const length = input.length;
  const prevPos = this.pos;

  let state = 1;
  let i = this.pos;

  while ( i < length ) {
    const char = this.codePointAt( i );
    const nextState = table[ state ]( char );
    if ( nextState ) {
      state = nextState;
      if ( char > 0xffff ) {
        i += 2;
      } else {
        i++;
      }
    } else {
      break;
    }
  }

  const id = finals[state];
  if ( id == null ) {
    this.unexpectedChar();
  }
  const image=input.slice(prevPos,i);
  const splitted=image.split(/\\\\r\\\\n?|\\\\n/g);
  const newLines=splitted.length-1;
  this.pos=i;
  if (newLines>0) {
    this._lineStart=this.pos-splitted[newLines].length;
    this._curLine+=newLines;
  }
  return {
    id,
    label: labels[id],
    image
  };
}
  }
class Parser extends Q.Parser{
constructor(text){super(new Tokenizer(text));}

    expect(id) {
      const token = this.token;
      if ( token.id !== id ) {
        const labels = this.tokenizer.labels;
        throw this.error(
          \`Unexpected token \${labels[token.id]||token.id}, expected \${labels[id]||id}\`
        );
      }
      this.next();
      return token;
    }
    
/*RULE2*/f2(){let $l=this.startNode();this.expect(2);loop:while(true){switch(this.token.id){case 2:this.expect(2);break;default:break loop;}}this.expect(2);return {type:\\"RULE2\\",loc:this.locNode($l)};}
/*RULE1*/f1(){let $l=this.startNode();loop:while(true){switch(this.token.id){case 1:this.expect(1);break;default:break loop;}}this.expect(1);this.f2();this.expect(\\"eof\\");return {type:\\"RULE1\\",loc:this.locNode($l)};}
parse(){return this.f1();}}
module.exports=Parser;
"
`;

exports[`supports empty: conflicts 1`] = `Array []`;

exports[`supports empty: generation 1`] = `
"/* eslint-disable */
const Q=require(\\"@quase/parser\\");


  class Tokenizer extends Q.Tokenizer{
    
    constructor(input){
      super(input);
      this.table=[c=>0,c=>c==65?2:0,c=>0];
      this.finals={\\"2\\":1};
      this.labels=[\\"\\",\\"'A'\\"];
    }
    initial(){return{id:\\"initial\\",label:\\"initial\\"};}
    eof(){return{id:\\"eof\\",label:\\"eof\\"};}
    
  codePointAt(index) {
    const first = this.input.charCodeAt(index);
    const size = this.input.length;
    if (
      first >= 0xD800 && first <= 0xDBFF &&
      size > index + 1
    ) {
      const second = this.input.charCodeAt(index + 1);
      if (second >= 0xDC00 && second <= 0xDFFF) {
        return (first - 0xD800) * 0x400 + second - 0xDC00 + 0x10000;
      }
    }
    return first;
  }
  
    
readToken() {
  const { input, table, finals, labels } = this;
  const length = input.length;
  const prevPos = this.pos;

  let state = 1;
  let i = this.pos;

  while ( i < length ) {
    const char = this.codePointAt( i );
    const nextState = table[ state ]( char );
    if ( nextState ) {
      state = nextState;
      if ( char > 0xffff ) {
        i += 2;
      } else {
        i++;
      }
    } else {
      break;
    }
  }

  const id = finals[state];
  if ( id == null ) {
    this.unexpectedChar();
  }
  const image=input.slice(prevPos,i);
  const splitted=image.split(/\\\\r\\\\n?|\\\\n/g);
  const newLines=splitted.length-1;
  this.pos=i;
  if (newLines>0) {
    this._lineStart=this.pos-splitted[newLines].length;
    this._curLine+=newLines;
  }
  return {
    id,
    label: labels[id],
    image
  };
}
  }
class Parser extends Q.Parser{
constructor(text){super(new Tokenizer(text));}

    expect(id) {
      const token = this.token;
      if ( token.id !== id ) {
        const labels = this.tokenizer.labels;
        throw this.error(
          \`Unexpected token \${labels[token.id]||token.id}, expected \${labels[id]||id}\`
        );
      }
      this.next();
      return token;
    }
    
/*RULE1*/f1(){let $l=this.startNode();switch(this.token.id){case 1:this.expect(1);break;default:}this.expect(\\"eof\\");return {type:\\"RULE1\\",loc:this.locNode($l)};}
parse(){return this.f1();}}
module.exports=Parser;
"
`;

exports[`typescript: generation 1`] = `
"/* eslint-disable */
import Q from \\"@quase/parser\\";
export type $Position = {pos:number;line:number;column:number;};
export type $Location = {start:$Position;end:$Position;};
export interface $Base<T> {type:T;loc:$Location;}
export type $5 = {id:5;label:\\"ID\\";image:string;};
export type $6 = {id:6;label:\\"NUM\\";image:string;};
export interface EXP extends $Base<'EXP'> {stuff:$6|$5;};
export interface PROGRAM extends $Base<'PROGRAM'> {params:($5)[];body:EXP;};

  class Tokenizer extends Q.Tokenizer{
    table:((c:number)=>number)[];finals:{[key:number]:number};labels:string[];
    constructor(input:string){
      super(input);
      this.table=[c=>0,c=>c==44?2:c==45?3:48<=c&&c<=57?5:65<=c&&c<=90?6:97<=c&&c<=100?6:c==101?7:c==102?10:103<=c&&c<=122?6:0,c=>0,c=>c==62?4:0,c=>0,c=>48<=c&&c<=57?5:0,c=>48<=c&&c<=57?6:65<=c&&c<=90?6:97<=c&&c<=122?6:0,c=>48<=c&&c<=57?6:65<=c&&c<=90?6:97<=c&&c<=109?6:c==110?8:111<=c&&c<=122?6:0,c=>48<=c&&c<=57?6:65<=c&&c<=90?6:97<=c&&c<=99?6:c==100?9:101<=c&&c<=122?6:0,c=>48<=c&&c<=57?6:65<=c&&c<=90?6:97<=c&&c<=122?6:0,c=>48<=c&&c<=57?6:65<=c&&c<=90?6:97<=c&&c<=116?6:c==117?11:118<=c&&c<=122?6:0,c=>48<=c&&c<=57?6:65<=c&&c<=90?6:97<=c&&c<=109?6:c==110?12:111<=c&&c<=122?6:0,c=>48<=c&&c<=57?6:65<=c&&c<=90?6:97<=c&&c<=122?6:0];
      this.finals={\\"2\\":2,\\"4\\":3,\\"5\\":6,\\"6\\":5,\\"7\\":5,\\"8\\":5,\\"9\\":4,\\"10\\":5,\\"11\\":5,\\"12\\":1};
      this.labels=[\\"\\",\\"'fun'\\",\\"','\\",\\"'->'\\",\\"'end'\\",\\"ID\\",\\"NUM\\"];
    }
    initial(){return{id:\\"initial\\",label:\\"initial\\"};}
    eof(){return{id:\\"eof\\",label:\\"eof\\"};}
    
  codePointAt(index:number):number {
    const first = this.input.charCodeAt(index);
    const size = this.input.length;
    if (
      first >= 0xD800 && first <= 0xDBFF &&
      size > index + 1
    ) {
      const second = this.input.charCodeAt(index + 1);
      if (second >= 0xDC00 && second <= 0xDFFF) {
        return (first - 0xD800) * 0x400 + second - 0xDC00 + 0x10000;
      }
    }
    return first;
  }
  
    
readToken() {
  const { input, table, finals, labels } = this;
  const length = input.length;
  const prevPos = this.pos;

  let state = 1;
  let i = this.pos;

  while ( i < length ) {
    const char = this.codePointAt( i );
    const nextState = table[ state ]( char );
    if ( nextState ) {
      state = nextState;
      if ( char > 0xffff ) {
        i += 2;
      } else {
        i++;
      }
    } else {
      break;
    }
  }

  const id = finals[state];
  if ( id == null ) {
    this.unexpectedChar();
  }
  const image=input.slice(prevPos,i);
  const splitted=image.split(/\\\\r\\\\n?|\\\\n/g);
  const newLines=splitted.length-1;
  this.pos=i;
  if (newLines>0) {
    this._lineStart=this.pos-splitted[newLines].length;
    this._curLine+=newLines;
  }
  return {
    id,
    label: labels[id],
    image
  };
}
  }
class Parser extends Q.Parser{
constructor(text:string){super(new Tokenizer(text));}

    expect(id:number|string) {
      const token = this.token;
      if ( token.id !== id ) {
        const labels = this.tokenizer.labels;
        throw this.error(
          \`Unexpected token \${labels[token.id]||token.id}, expected \${labels[id]||id}\`
        );
      }
      this.next();
      return token;
    }
    
/*EXP*/f2():EXP{let $l=this.startNode();let $n_stuff;switch(this.token.id){case 6:$n_stuff=this.expect(6);break;case 5:$n_stuff=this.expect(5);break;default:this.unexpected();}return {type:\\"EXP\\",stuff:$n_stuff,loc:this.locNode($l)};}
/*PROGRAM*/f1():PROGRAM{let $l=this.startNode();let $n_params=[],$n_body;this.expect(1);switch(this.token.id){case 5:$n_params.push(this.expect(5));loop:while(true){switch(this.token.id){case 2:this.expect(2);$n_params.push(this.expect(5));break;default:break loop;}}break;}this.expect(3);$n_body=this.f2();this.expect(4);this.expect(\\"eof\\");return {type:\\"PROGRAM\\",params:$n_params,body:$n_body,loc:this.locNode($l)};}
parse(){return this.f1();}}
export default Parser;
"
`;
